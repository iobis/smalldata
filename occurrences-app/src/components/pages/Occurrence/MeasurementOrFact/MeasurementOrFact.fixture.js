import MeasurementOrFact from './MeasurementOrFact'
import { faAngleDown } from '@fortawesome/free-solid-svg-icons'
import { library } from '@fortawesome/fontawesome-svg-core'

library.add(faAngleDown)

export default [{
  component: MeasurementOrFact,
  name:      'for empty data',
  props:     {
    data:     getEmptyData(),
    onChange: console.log
  }
}]

export function getEmptyData() {
  return []
}

export function getData() {
  return [{
    type:  'general-measurement-1',
    unit:  'general-measurement-1-unit-1',
    value: '123',
    units: getPressureUnits()
  }]
}

export function getPressureUnits() {
  return [
    { id: 'http://general-measurement-1-unit-1/', name: 'general-measurement-1-unit-1' },
    { id: 'http://general-measurement-1-unit-2/', name: 'general-measurement-1-unit-2' }
  ]
}
