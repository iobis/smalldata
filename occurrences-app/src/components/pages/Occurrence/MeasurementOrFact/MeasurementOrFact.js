import CopyPreviousData from '../CopyPreviousData'
import Dropdown from '../../../form/Dropdown'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { getGeneralMeasurements, getSpecificMeasurements } from '../../../../clients/measurments'
import { useTranslation } from 'react-i18next'

export default function MeasurementOrFact({ onChange }) {
  const { t } = useTranslation()
  const generalMeasurements = getGeneralMeasurements()
  const specificMeasurements = getSpecificMeasurements()
  const [suppliedMeasurements, setSuppliedMeasurements] = useState([])

  function addSuppliedMeasurements(measurment) {
    const updatedMeasurements = [...suppliedMeasurements, measurment]
      .sort((left, right) => left.type.toLowerCase().localeCompare(right.type.toLowerCase()))
    setSuppliedMeasurements(updatedMeasurements)
  }

  function removeSuppliedMeasurement(index) {
    const updatedMeasurements = suppliedMeasurements.filter((_, i) => i !== index)
      .sort((left, right) => left.type.toLowerCase().localeCompare(right.type.toLowerCase()))
    setSuppliedMeasurements(updatedMeasurements)
  }

  return (
    <div className="measurement-or-fact section is-fluid">
      <div className="columns">
        <div className="column is-half">
          <h1 className="title">{t('occurrenceForm.measurementOrFact.general.title')}</h1>
          <h2 className="subtitle">{t('occurrenceForm.measurementOrFact.general.subtitle')}</h2>
          <table className="general table is-fullwidth is-striped is-hoverable">
            <thead>
            <tr>
              <th style={{ width: '60%' }}>type</th>
              <th style={{ width: '20%' }}>unit</th>
              <th style={{ width: '10%' }}>value</th>
              <th style={{ width: '10%' }}/>
            </tr>
            </thead>
            <tbody>
            {generalMeasurements.map(generalMeasurement => (
              <MeasurementRow
                key={generalMeasurement.type}
                onClickAdd={addSuppliedMeasurements}
                type={generalMeasurement.type}
                units={generalMeasurement.units}/>
            ))}
            </tbody>
          </table>
          <h1 className="title">{t('occurrenceForm.measurementOrFact.specific.title')}</h1>
          <h2 className="subtitle">{t('occurrenceForm.measurementOrFact.specific.subtitle')}</h2>
          <table className="specific table is-fullwidth is-striped is-hoverable">
            <thead>
            <tr>
              <th style={{ width: '60%' }}>type</th>
              <th style={{ width: '20%' }}>unit</th>
              <th style={{ width: '10%' }}>value</th>
              <th style={{ width: '10%' }}/>
            </tr>
            </thead>
            <tbody>
            {specificMeasurements.map(specificMeasurement => (
              <MeasurementRow
                key={specificMeasurement.type}
                onClickAdd={addSuppliedMeasurements}
                type={specificMeasurement.type}
                units={specificMeasurement.units}/>
            ))}
            </tbody>
          </table>
        </div>
        <div className="column is-half supplied">
          <h1 className="title">{t('occurrenceForm.measurementOrFact.supplied.title')}</h1>
          <h2 className="subtitle">&nbsp;</h2>
          <table className="supplied table is-fullwidth is-striped is-hoverable">
            <thead>
            <tr>
              <th style={{ width: '60%' }}>type</th>
              <th style={{ width: '20%' }}>unit</th>
              <th style={{ width: '10%' }}>value</th>
              <th style={{ width: '10%' }}/>
            </tr>
            </thead>
            <tbody>
            {suppliedMeasurements.map(({ type, unit, value }, index) => (
              // eslint-disable-next-line react/no-array-index-key
              <tr className="fieldrow" key={type + unit + value + index}>
                <td>{type}</td>
                <td>{unit}</td>
                <td>{value}</td>
                <td><a className="remove button" onClick={() => removeSuppliedMeasurement(index)}>remove</a></td>
              </tr>
            ))}
            </tbody>
          </table>
        </div>
      </div>
      <CopyPreviousData/>
    </div>
  )
}

MeasurementOrFact.propTypes = {
  onChange: PropTypes.func.isRequired
}

function MeasurementRow({ onClickAdd, type, units }) {
  const [selectedUnit, setSelectedUnit] = useState(units[0].name)
  const [selectedValue, setSelectedValue] = useState('')

  return (
    <tr className="measurement-row fieldrow">
      <td>{type}</td>
      <td>
        <Dropdown
          onChange={(value) => setSelectedUnit(value)}
          options={units.map(unit => unit.name)}
          value={selectedUnit}/>
      </td>
      <td>
        <input
          className="input"
          onChange={(e) => setSelectedValue(e.target.value)}
          type="text"
          value={selectedValue}/>
      </td>
      <td>
        <a
          className="add button"
          onClick={() => onClickAdd({ unit: selectedUnit, type, value: selectedValue })}>
          add
        </a>
      </td>
    </tr>
  )
}

MeasurementRow.propTypes = {
  onClickAdd: PropTypes.func.isRequired,
  type:       PropTypes.string.isRequired,
  units:      PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired
  })).isRequired
}
