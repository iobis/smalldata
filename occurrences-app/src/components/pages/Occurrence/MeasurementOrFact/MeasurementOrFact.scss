.measurement-or-fact {
  margin-bottom: 1.25rem;

  .supplied {
    background-color: $background;
  }

  td {
    vertical-align: middle;
  }
}
