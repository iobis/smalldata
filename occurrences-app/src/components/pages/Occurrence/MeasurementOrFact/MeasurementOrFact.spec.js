import MeasurementOrFact from './MeasurementOrFact'
import React from 'react'
import { getEmptyData } from './MeasurementOrFact.fixture'
import { mount } from 'enzyme'

describe('MeasurementOrFact', () => {
  let wrapper

  it('renders correctly', () => {
    expect(mount(createComponent())).toMatchSnapshot()
  })

  describe('when adding measurement with default unit', () => {
    beforeAll(() => {
      wrapper = mount(createComponent())
    })

    it('does not render any supplied measurements', () => {
      expect(wrapper.find('.supplied .fieldrow')).toHaveLength(0)
    })

    describe('and then changing unit and adding measurement', () => {
      beforeAll(() => {
        wrapper.find('.general .measurement-row .input').at(0).simulate('change', { target: { value: '10' } })
        wrapper.find('.general .measurement-row .button.add').at(0).simulate('click')
      })

      it('renders supplied value with default unit', () => {
        expect(wrapper.find('.supplied .fieldrow')).toHaveLength(1)
        expect(wrapper.find('.supplied .fieldrow td').at(0).text()).toBe('Pressure')
        expect(wrapper.find('.supplied .fieldrow td').at(1).text()).toBe('Kilogram')
        expect(wrapper.find('.supplied .fieldrow td').at(2).text()).toBe('10')
      })
    })
  })

  describe('when adding measurement with updated unit', () => {
    beforeAll(() => {
      wrapper = mount(createComponent())
    })

    it('does not render any supplied measurements', () => {
      expect(wrapper.find('.supplied .fieldrow')).toHaveLength(0)
    })

    describe('and then changing unit and adding measurement', () => {
      beforeAll(() => {
        wrapper.find('.general .measurement-row .dropdown').at(0).simulate('click')
        wrapper.find('.general .measurement-row .dropdown-item').at(1).simulate('click')
        wrapper.find('.general .measurement-row .input').at(0).simulate('change', { target: { value: '20' } })
        wrapper.find('.general .measurement-row .button.add').at(0).simulate('click')
      })

      it('renders supplied value with selected unit', () => {
        expect(wrapper.find('.supplied .fieldrow')).toHaveLength(1)
        expect(wrapper.find('.supplied .fieldrow td').at(0).text()).toBe('Pressure')
        expect(wrapper.find('.supplied .fieldrow td').at(1).text()).toBe('Gram')
        expect(wrapper.find('.supplied .fieldrow td').at(2).text()).toBe('20')
      })
    })
  })

  describe('when adding multiple measurements', () => {
    beforeAll(() => {
      wrapper = mount(createComponent())
      wrapper.find('.general .measurement-row .input').at(0).simulate('change', { target: { value: '10' } })
      wrapper.find('.general .measurement-row .button.add').at(0).simulate('click')
      wrapper.find('.general .measurement-row .input').at(0).simulate('change', { target: { value: '20' } })
      wrapper.find('.general .measurement-row .button.add').at(0).simulate('click')
      wrapper.find('.general .measurement-row .input').at(0).simulate('change', { target: { value: '30' } })
      wrapper.find('.general .measurement-row .button.add').at(0).simulate('click')
    })

    it('renders 3 measurements', () => {
      expect(wrapper.find('.supplied .fieldrow')).toHaveLength(3)
      expect(wrapper.find('.supplied .fieldrow').map(el => el.find('td').at(2).text())).toEqual(['10', '20', '30'])
    })

    describe('and then removing first one', () => {
      beforeAll(() => {
        wrapper.find('.supplied .button.remove').first().simulate('click')
      })

      it('renders 2 measurements', () => {
        expect(wrapper.find('.supplied .fieldrow')).toHaveLength(2)
        expect(wrapper.find('.supplied .fieldrow').map(el => el.find('td').at(2).text())).toEqual(['20', '30'])
      })

      describe('and then removing last one', () => {
        beforeAll(() => {
          wrapper.find('.supplied .button.remove').last().simulate('click')
        })

        it('renders 1 measurements', () => {
          expect(wrapper.find('.supplied .fieldrow')).toHaveLength(1)
          expect(wrapper.find('.supplied .fieldrow').map(el => el.find('td').at(2).text())).toEqual(['20'])
        })
      })
    })
  })
})

function createComponent(props) {
  const defaultProps = {
    data:     getEmptyData(),
    onChange: jest.fn()
  }
  return <MeasurementOrFact {...defaultProps} {...props}/>
}
