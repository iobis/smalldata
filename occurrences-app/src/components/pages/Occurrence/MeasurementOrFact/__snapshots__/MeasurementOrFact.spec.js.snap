// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasurementOrFact renders correctly 1`] = `
<MeasurementOrFact
  data={Array []}
  onChange={[MockFunction]}
>
  <div
    className="measurement-or-fact section is-fluid"
  >
    <div
      className="columns"
    >
      <div
        className="column is-half"
      >
        <h1
          className="title"
        >
          occurrenceForm.measurementOrFact.general.title
        </h1>
        <h2
          className="subtitle"
        >
          occurrenceForm.measurementOrFact.general.subtitle
        </h2>
        <table
          className="general table is-fullwidth is-striped is-hoverable"
        >
          <thead>
            <tr>
              <th
                style={
                  Object {
                    "width": "60%",
                  }
                }
              >
                common.type
              </th>
              <th
                style={
                  Object {
                    "width": "20%",
                  }
                }
              >
                common.unit
              </th>
              <th
                style={
                  Object {
                    "width": "10%",
                  }
                }
              >
                common.value
              </th>
              <th
                style={
                  Object {
                    "width": "10%",
                  }
                }
              />
            </tr>
          </thead>
          <tbody>
            <MeasurementRow
              key="general-measurement-1"
              onClickAdd={[Function]}
              type="general-measurement-1"
              units={
                Array [
                  Object {
                    "id": "http://general-measurement-1-unit-1/",
                    "name": "general-measurement-1-unit-1",
                  },
                  Object {
                    "id": "http://general-measurement-1-unit-2/",
                    "name": "general-measurement-1-unit-2",
                  },
                ]
              }
            >
              <tr
                className="measurement-row fieldrow"
              >
                <td>
                  general-measurement-1
                </td>
                <td>
                  <Dropdown
                    onChange={[Function]}
                    options={
                      Array [
                        "general-measurement-1-unit-1",
                        "general-measurement-1-unit-2",
                      ]
                    }
                    value="general-measurement-1-unit-1"
                  >
                    <div
                      className="dropdown"
                      onClick={[Function]}
                    >
                      <div
                        className="dropdown-trigger"
                      >
                        <button
                          aria-controls="dropdown-menu"
                          aria-haspopup="true"
                          className="button"
                        >
                          <span
                            className="selected-value"
                          >
                            general-measurement-1-unit-1
                          </span>
                          <span
                            className="icon is-small"
                          >
                            <FontAwesomeIcon
                              className="angle-down"
                              icon="angle-down"
                            >
                              <i
                                className="angle-down"
                                icon="angle-down"
                              />
                            </FontAwesomeIcon>
                          </span>
                        </button>
                      </div>
                      <div
                        className="dropdown-menu"
                        id="dropdown-menu"
                        role="menu"
                      >
                        <div
                          className="dropdown-content"
                        >
                          <div
                            className="dropdown-item is-active"
                            key="general-measurement-1-unit-1"
                            onClick={[Function]}
                          >
                            general-measurement-1-unit-1
                          </div>
                          <div
                            className="dropdown-item"
                            key="general-measurement-1-unit-2"
                            onClick={[Function]}
                          >
                            general-measurement-1-unit-2
                          </div>
                        </div>
                      </div>
                    </div>
                  </Dropdown>
                </td>
                <td>
                  <input
                    className="input"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <a
                    className="add button"
                    onClick={[Function]}
                  >
                    common.add
                  </a>
                </td>
              </tr>
            </MeasurementRow>
            <MeasurementRow
              key="general-measurement-2"
              onClickAdd={[Function]}
              type="general-measurement-2"
              units={
                Array [
                  Object {
                    "id": "general-measurement-2-unit-1",
                    "name": "general-measurement-2-unit-1",
                  },
                ]
              }
            >
              <tr
                className="measurement-row fieldrow"
              >
                <td>
                  general-measurement-2
                </td>
                <td>
                  <Dropdown
                    onChange={[Function]}
                    options={
                      Array [
                        "general-measurement-2-unit-1",
                      ]
                    }
                    value="general-measurement-2-unit-1"
                  >
                    <div
                      className="dropdown"
                      onClick={[Function]}
                    >
                      <div
                        className="dropdown-trigger"
                      >
                        <button
                          aria-controls="dropdown-menu"
                          aria-haspopup="true"
                          className="button"
                        >
                          <span
                            className="selected-value"
                          >
                            general-measurement-2-unit-1
                          </span>
                          <span
                            className="icon is-small"
                          >
                            <FontAwesomeIcon
                              className="angle-down"
                              icon="angle-down"
                            >
                              <i
                                className="angle-down"
                                icon="angle-down"
                              />
                            </FontAwesomeIcon>
                          </span>
                        </button>
                      </div>
                      <div
                        className="dropdown-menu"
                        id="dropdown-menu"
                        role="menu"
                      >
                        <div
                          className="dropdown-content"
                        >
                          <div
                            className="dropdown-item is-active"
                            key="general-measurement-2-unit-1"
                            onClick={[Function]}
                          >
                            general-measurement-2-unit-1
                          </div>
                        </div>
                      </div>
                    </div>
                  </Dropdown>
                </td>
                <td>
                  <input
                    className="input"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <a
                    className="add button"
                    onClick={[Function]}
                  >
                    common.add
                  </a>
                </td>
              </tr>
            </MeasurementRow>
            <MeasurementRow
              key="general-measurement-3"
              onClickAdd={[Function]}
              type="general-measurement-3"
              units={
                Array [
                  Object {
                    "id": "general-measurement-3-unit-1",
                    "name": "general-measurement-3-unit-1",
                  },
                ]
              }
            >
              <tr
                className="measurement-row fieldrow"
              >
                <td>
                  general-measurement-3
                </td>
                <td>
                  <Dropdown
                    onChange={[Function]}
                    options={
                      Array [
                        "general-measurement-3-unit-1",
                      ]
                    }
                    value="general-measurement-3-unit-1"
                  >
                    <div
                      className="dropdown"
                      onClick={[Function]}
                    >
                      <div
                        className="dropdown-trigger"
                      >
                        <button
                          aria-controls="dropdown-menu"
                          aria-haspopup="true"
                          className="button"
                        >
                          <span
                            className="selected-value"
                          >
                            general-measurement-3-unit-1
                          </span>
                          <span
                            className="icon is-small"
                          >
                            <FontAwesomeIcon
                              className="angle-down"
                              icon="angle-down"
                            >
                              <i
                                className="angle-down"
                                icon="angle-down"
                              />
                            </FontAwesomeIcon>
                          </span>
                        </button>
                      </div>
                      <div
                        className="dropdown-menu"
                        id="dropdown-menu"
                        role="menu"
                      >
                        <div
                          className="dropdown-content"
                        >
                          <div
                            className="dropdown-item is-active"
                            key="general-measurement-3-unit-1"
                            onClick={[Function]}
                          >
                            general-measurement-3-unit-1
                          </div>
                        </div>
                      </div>
                    </div>
                  </Dropdown>
                </td>
                <td>
                  <input
                    className="input"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <a
                    className="add button"
                    onClick={[Function]}
                  >
                    common.add
                  </a>
                </td>
              </tr>
            </MeasurementRow>
          </tbody>
        </table>
        <h1
          className="title"
        >
          occurrenceForm.measurementOrFact.specific.title
        </h1>
        <h2
          className="subtitle"
        >
          occurrenceForm.measurementOrFact.specific.subtitle
        </h2>
        <table
          className="specific table is-fullwidth is-striped is-hoverable"
        >
          <thead>
            <tr>
              <th
                style={
                  Object {
                    "width": "60%",
                  }
                }
              >
                common.type
              </th>
              <th
                style={
                  Object {
                    "width": "20%",
                  }
                }
              >
                common.unit
              </th>
              <th
                style={
                  Object {
                    "width": "10%",
                  }
                }
              >
                common.value
              </th>
              <th
                style={
                  Object {
                    "width": "10%",
                  }
                }
              />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        className="column is-half supplied"
      >
        <h1
          className="title"
        >
          occurrenceForm.measurementOrFact.supplied.title
        </h1>
        <h2
          className="subtitle"
        >
           
        </h2>
        <table
          className="supplied table is-fullwidth is-striped is-hoverable"
        >
          <thead>
            <tr>
              <th
                style={
                  Object {
                    "width": "60%",
                  }
                }
              >
                common.type
              </th>
              <th
                style={
                  Object {
                    "width": "20%",
                  }
                }
              >
                common.unit
              </th>
              <th
                style={
                  Object {
                    "width": "10%",
                  }
                }
              >
                common.value
              </th>
              <th
                style={
                  Object {
                    "width": "5%",
                  }
                }
              />
              <th
                style={
                  Object {
                    "width": "5%",
                  }
                }
              />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
    <CopyPreviousData>
      <div
        className="columns"
      >
        <div
          className="copy-previous-data column is-narrow"
        >
          copy data from previous entry
        </div>
      </div>
    </CopyPreviousData>
  </div>
</MeasurementOrFact>
`;
